llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -3080,22 +3108,31 @@ bool SimplifyCFGOpt::speculativelyExecuteBB(BranchInst *BI,
 
     // Only speculatively execute a single instruction (not counting the
     // terminator) for now.
-    ++SpeculatedInstructions;
+    bool IsSafeCheapLoadStore =
+        HositLoadsStores && isSafeCheapLoadStore(&I, TTI);
+    // Not count load/store into cost if target supports conditional faulting
+    // b/c it's cheap to speculate it.
```

Comment:
At least as written, I don't think this reasoning makes a lot of sense. It's not like speculating a cload/cstore is *cheaper* than speculating an add, right?

And just ignoring the number entirely surely cannot be right. If you have a conditional block with 1000 stores, it does not make sense to convert all of them into cstore. I think your current implementation is going to do that?

Reply:
Speculating a cload/cstore should be cheaper than an add b/c it behaves as a nop when the condition is not satisfied. We can add a flag to control the threshold of  speculated cload/cstore (let's use 6 as default FTTB).

Reply:
Correct myself. Speculating a cload/cstore is not cheaper than speculating an add. But the load latency is typically 4, if we can load earlier, the consumer of the load will not have to wait for 4 cycles and hide the pipeline.